### DISCUSSION TITLE
Clean way to add docker-desktop cask as a dependency

### DISCUSSION THREAD

**Discussion by tonytrg** [Homebrew/discussions#3163] [Casks]
Hello community 👋 ,

I would like to add docker-desktop as a [dependency](https://formulae.brew.sh/cask/docker).
in the style of `depends_on "docker"`, right now it would simply install that non cask [dependency](https://formulae.brew.sh/formula/docker).

What is the best way to implement the cask as a dependency?
Other idea I have it to maybe use a `post-install`, which seems kinda hacky to me.

Best regards, Tony

**Comment by SMillerDev** [comment]
Is it a dependency in a cask or formula?

**Reply by tonytrg** [reply]
It exists as both, a formula and a cask.
But in my case I want the cask as dependency, and I am not sure if brew allows that.
 This is the cask I am talking about. [Docker-desktop-cask](https://formulae.brew.sh/cask/docker).
 And I have a private formula, where I would like this cask as a dependency á la `depends_on docker #the cask version`

**Reply by SMillerDev** [reply]
So `depends_on "homebrew/cask/docker-desktop` should work I think.

**Reply by tonytrg** [reply]
i tried some different variations which unfortunately did not work:
- `depends_on "homebrew/cask/docker-desktop"`
- `depends_on cask: "homebrew/cask/docker-desktop"`
- `depends_on cask "homebrew/cask/docker-desktop"`

In the cookbook under the chaption [Specifying other formulae as dependencies](https://docs.brew.sh/Formula-Cookbook#specifying-other-formulae-as-dependencies) it states this: 

> A Symbol (e.g. :xcode) specifies a [Requirement](https://rubydoc.brew.sh/Requirement) which can be fulfilled by one or more formulae, casks or other system-wide installed software (e.g. Xcode).

The docs seem a little bit unclear to me in this aspect. I also wonder if adding casks as dependencies should not be a more popular problem, the research I have done is conflicting. I don't see a definitive yes or no on this question.

**Reply by tonytrg** [reply]
This is the output:
`brew install private-formula
Running `brew update --preinstall`...
==> Auto-updated Homebrew!
Updated Homebrew from be8385785 to 9c58f9a77.
No changes to formulae.

Warning: No available formula with the name "homebrew/cask/docker-desktop" (dependency of private-formula).
==> Searching for similarly named formulae...
Error: No similarly named formulae found.
==> Searching for a previously deleted formula (in the last month)...
Error: No previously deleted formula found.`

**Comment by tlk** [comment]
> What is the best way to implement the cask as a dependency?

The output from `grep -r -e 'depends.*cask' -e 'depends.*formula' $(brew --repository)/Library/Taps/homebrew/` seems to indicate that a cask can depend on casks and formulas, but a formula can only depend on formulas (not casks).

Here is an example of a cask that depends on the docker cask:
https://github.com/Homebrew/homebrew-cask/blob/06a31f6a264c647603193eefac0cb9e66b36df18/Casks/lando.rb#L23

**Reply by SMillerDev** [reply]
That's actually just a homebrew/core rule. As far as I know it's entirely possible to do this (because we used to in core).

**Comment by Bo98** [comment]
Depending on casks is not possible directly.

This is because Casks are still seen as "external" dependencies - they effectively just run upstream installers, so there's no real difference between a cask installation and another external installation.

What you can do is have custom requirements that inform users to install the cask first. We had a couple of them ourselves in the past when we were more lenient about Java and macFUSE: https://github.com/Homebrew/brew/blob/3.0.11/Library/Homebrew/extend/os/mac/requirements/osxfuse_requirement.rb. This will also allow people who have Docker Desktop installed outside of Homebrew/cask to still use that.

Custom requirements are embeddable inside formulae: see `glibc.rb` in Homebrew/core for an example, which has special requirements for bootstrapping reasons.

**Comment by tonytrg** [comment]
Thank you all for the answers.
I went with the option to add manual steps in the caveat. Since after running the formula, these are only 2-3 optional steps I did not want to create a complicated workflow.

Thank you all for the answers!

**Comment by gibfahn** [comment]
Just cross-linking to other times this has been discussed: https://github.com/Homebrew/brew/issues/17326, https://github.com/orgs/Homebrew/discussions/5788
